tokio/src/process/unix/pidfd_reaper.rs

```diff
@@ -0,0 +1,313 @@
+        if let Ok(Some(_)) = orphan.try_wait() {
+            return;
+        }
+
+        self.orphan_queue.push_orphan(orphan);
+    }
+}
+
+#[cfg(all(test, not(loom), not(miri)))]
+mod test {
+    use super::*;
+    use crate::{
+        process::unix::orphan::test::MockQueue,
+        runtime::{Builder as RuntimeBuilder, Runtime},
+    };
+    use std::process::{Command, Output};
+
+    fn create_runtime() -> Runtime {
+        RuntimeBuilder::new_current_thread()
+            .enable_io()
+            .build()
+            .unwrap()
+    }
+
+    fn run_test(fut: impl Future<Output = ()>) {
+        create_runtime().block_on(fut)
+    }
+
+    fn is_pidfd_available() -> bool {
+        let Output { stdout, status, .. } = Command::new("uname").arg("-r").output().unwrap();
+        assert!(status.success());
+        let stdout = String::from_utf8_lossy(&stdout);
+
+        let mut kernel_version_iter = stdout.split_once('-').unwrap().0.split('.');
+        let major: u32 = kernel_version_iter.next().unwrap().parse().unwrap();
+        let minor: u32 = kernel_version_iter.next().unwrap().parse().unwrap();
+
+        major >= 6 || (major == 5 && minor >= 10)
+    }
+
+    #[test]
+    fn test_pidfd_reaper_poll() {
+        if !is_pidfd_available() {
+            eprintln!("pidfd is not available on this linux kernel, skip this test");
+            return;
+        }
+
+        let queue = MockQueue::new();
+
+        run_test(async {
+            let child = Command::new("true").spawn().unwrap();
+            let pidfd_reaper = PidfdReaper::new(child, &queue).unwrap();
+
+            let exit_status = pidfd_reaper.await.unwrap();
+            assert!(exit_status.success());
+        });
+
+        assert!(queue.all_enqueued.borrow().is_empty());
+    }
+
+    #[test]
+    fn test_pidfd_reaper_kill() {
+        if !is_pidfd_available() {
+            eprintln!("pidfd is not available on this linux kernel, skip this test");
+            return;
+        }
+
+        let queue = MockQueue::new();
+
+        run_test(async {
+            let child = Command::new("sleep").arg("1800").spawn().unwrap();
+            let mut pidfd_reaper = PidfdReaper::new(child, &queue).unwrap();
+
+            pidfd_reaper.kill().unwrap();
+
+            let exit_status = pidfd_reaper.await.unwrap();
+            assert!(!exit_status.success());
+        });
+
+        assert!(queue.all_enqueued.borrow().is_empty());
+    }
+
+    #[test]
+    fn test_pidfd_reaper_drop() {
+        if !is_pidfd_available() {
+            eprintln!("pidfd is not available on this linux kernel, skip this test");
+            return;
+        }
+
+        let queue = MockQueue::new();
+
+        run_test(async {
+            let child = Command::new("sleep").arg("1800").spawn().unwrap();
+            let _pidfd_reaper = PidfdReaper::new(child, &queue).unwrap();
+        });
+
+        assert_eq!(queue.all_enqueued.borrow().len(), 1);
+    }
```

Comment:
How long does this test take to run? Does it leave a call to `sleep` running in the background after it passes?

Can we include some sort of kill-on-drop at the end?

Reply:
According to the [CI log](https://github.com/tokio-rs/tokio/actions/runs/7481388342/job/20362976681?pr=6152#step:6:115) it takes 0.11s and leaks the `sleep` process.

I think it's necessary to add a kill-on-drop logic at the end.

Reply:
I've pushed a fix for this and verified the [CI log](https://github.com/tokio-rs/tokio/actions/runs/7488884459/job/20384335032?pr=6152#step:6:69) that it is indeed fixed.